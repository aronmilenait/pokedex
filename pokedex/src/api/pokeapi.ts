import axios from "axios";

const pokeApi = axios.create({
  baseURL: "https://pokeapi.co/api/v2",
});

export const getPokemonList = async (limit: number, offset: number) => {
  try {
    const response = await pokeApi.get(
      `/pokemon?limit=${limit}&offset=${offset}`
    );
    return response.data;
  } catch (error) {
    console.error(error);
  }
};

export const getPokemonDetails = async (id: number) => {
  try {
    const response = await pokeApi.get(`/pokemon/${id}`);
    return response.data;
  } catch (error) {
    console.error(error);
  }
};

export default pokeApi;
