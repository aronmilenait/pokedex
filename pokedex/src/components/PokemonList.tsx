import { useEffect, useState } from "react";
import { getPokemonList } from "../api/pokeapi";
import PokemonCard from "./PokemonCard";

interface Pokemon {
  name: string;
  id: number;
  image: string;
}

const PokemonList = () => {
  const [pokemonList, setPokemonList] = useState<Pokemon[]>([]);
  const [currentPage, setCurrentPage] = useState(1);
  const pageSize = 24;
  const [totalPages, setTotalPages] = useState(0);

  useEffect(() => {
    const fetchPokemonList = async () => {
      try {
        const response = await getPokemonList(
          pageSize,
          (currentPage - 1) * pageSize
        );
        setPokemonList(response.results);

        setTotalPages(Math.ceil(response.count / pageSize));

        const pokemonDetails = response.results.map(
          (pokemon: { name: string; url: string }) => {
            const id = parseInt(pokemon.url.split("/")[6]);
            const imageUrl = `https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/${id}.png`;

            return {
              name: pokemon.name,
              id: id,
              image: imageUrl,
            };
          }
        );

        setPokemonList(pokemonDetails);
      } catch (error) {
        console.error("Error fetching Pokemon list:", error);
      }
    };

    fetchPokemonList();
  }, [currentPage]);

  const nextPage = () => {
    setCurrentPage(currentPage + 1);
  };

  const prevPage = () => {
    setCurrentPage(currentPage - 1);
  };

  return (
    <>
      <div
        id="pagination"
        className="flex justify-center items-center p-5 bg-very-light-blue"
      >
        <button
          onClick={prevPage}
          disabled={currentPage === 1}
          className="bg-blue hover:bg-dark-blue text-white font-bold py-2 px-4 mr-2 rounded focus:outline-none focus:shadow-outline"
        >
          Previous
        </button>
        <p className="mr-2 bg-blue text-white font-bold py-2 px-4 rounded">
          Page {currentPage}
        </p>
        <button
          onClick={nextPage}
          disabled={currentPage === totalPages}
          className="bg-blue hover:bg-dark-blue text-white font-bold py-2 px-4 rounded focus:outline-none focus:shadow-outline"
        >
          Next
        </button>
      </div>
      <section
        id="pokemon-list"
        className="grid grid-cols-3 md:grid-cols-6 gap-2 bg-very-light-blue p-10"
      >
        {pokemonList.map((pokemon) => (
          <PokemonCard key={pokemon.name} pokemon={pokemon} />
        ))}
      </section>
    </>
  );
};

export default PokemonList;
