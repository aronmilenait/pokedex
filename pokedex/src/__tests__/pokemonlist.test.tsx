import { render, waitFor } from '@testing-library/react';
import PokemonList from '../components/PokemonList';
import { getPokemonList } from '../api/pokeapi';

jest.mock('../api/pokeapi');

describe('PokemonList component', () => {
    beforeEach(() => {
        jest.resetAllMocks();
    });

    test('renders PokemonList', async () => {
        const mockResponse = {
            results: [
                { name: 'bulbasaur', url: 'https://pokeapi.co/api/v2/pokemon/1/' },
                { name: 'ivysaur', url: 'https://pokeapi.co/api/v2/pokemon/2/' },
                { name: 'venusaur', url: 'https://pokeapi.co/api/v2/pokemon/3/' }
            ],
            count: 3
        };

        (getPokemonList as jest.Mock).mockResolvedValue(mockResponse);

        const { container } = render(<PokemonList />);

        await waitFor(() => expect(getPokemonList).toHaveBeenCalledTimes(1));

        expect(container).toMatchSnapshot();
    });
});
