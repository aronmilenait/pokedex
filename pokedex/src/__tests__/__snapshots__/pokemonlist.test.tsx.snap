// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonList component renders PokemonList 1`] = `
<div>
  <div
    class="flex justify-center items-center p-5 bg-very-light-blue"
    id="pagination"
  >
    <button
      class="bg-blue hover:bg-dark-blue text-white font-bold py-2 px-4 mr-2 rounded focus:outline-none focus:shadow-outline"
      disabled=""
    >
      Previous
    </button>
    <p
      class="mr-2 bg-blue text-white font-bold py-2 px-4 rounded"
    >
      Page 
      1
    </p>
    <button
      class="bg-blue hover:bg-dark-blue text-white font-bold py-2 px-4 rounded focus:outline-none focus:shadow-outline"
      disabled=""
    >
      Next
    </button>
  </div>
  <section
    class="grid grid-cols-3 md:grid-cols-6 gap-2 bg-very-light-blue p-10"
    id="pokemon-list"
  >
    <div
      class="bg-light-blue p-4 text-center shadow-md"
    >
      <img
        alt="bulbasaur"
        class="w-24 mx-auto mb-4 hover:scale-125 transition-transform duration-200 ease-in-out"
        src="https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/1.png"
      />
      <p
        class="text-lg font-bold text-white text-shadow"
      >
        bulbasaur
      </p>
    </div>
    <div
      class="bg-light-blue p-4 text-center shadow-md"
    >
      <img
        alt="ivysaur"
        class="w-24 mx-auto mb-4 hover:scale-125 transition-transform duration-200 ease-in-out"
        src="https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/2.png"
      />
      <p
        class="text-lg font-bold text-white text-shadow"
      >
        ivysaur
      </p>
    </div>
    <div
      class="bg-light-blue p-4 text-center shadow-md"
    >
      <img
        alt="venusaur"
        class="w-24 mx-auto mb-4 hover:scale-125 transition-transform duration-200 ease-in-out"
        src="https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/3.png"
      />
      <p
        class="text-lg font-bold text-white text-shadow"
      >
        venusaur
      </p>
    </div>
  </section>
</div>
`;
